arkouda/groupbyclass.py

```diff
@@ -970,6 +975,333 @@ def broadcast(self, values : pdarray, permute : bool=True) -> pdarray:
         repMsg = generic_msg(cmd=cmd,args=args)
         return create_pdarray(repMsg)
 
+    @staticmethod
+    def build_from_components(keys, permutation, user_defined_name: str = None) -> GroupBy:
+        """
+        function to build a new GroupBy object from component keys and permutation.
+
+        Parameters
+        ----------
+            keys : (list of) pdarray, Strings, or Categorical
+            permutation : pdarray
+            user_defined_name : str (Optional) Passing a name will init the new GroupBy
+                and assign it the given name
+
+        Returns
+        -------
+        GroupBy
+            The GroupBy object created by using the given components
+
+        """
+        if len(keys) == 1:
+            g = GroupBy(keys[0])
+        else:
+            g = GroupBy(keys)
+
+        g.permutation = permutation
+        g.name = user_defined_name
+
+        print(g.keys)
+        return g
```

Comment:
To avoid the expensive recomputation, you could do something like what's done in `arkouda/categorical.py:Categorical.__init__`, where the presence of `**kwargs` is used to determine the initialization path, e.g. when called from `Categorical.from_codes()`, which is analogous to your `GroupBy.build_from_components()` function. 

That said, I just submitted a PR (#1330 ) that completely refactors the `GroupBy.__init__` function, so I might wait until that PR gets resolved before making your own modifications to `__init__`.

Reply:
I pushed up Ethan's idea for this this morning, but I'll take a look at how Categorical is doing it. I'm going to convert this PR into a draft until #1330 gets merged. Then I'll revisit this and make sure there are no conflicts before setting it back to active